cmd/gossamer/account.go

```diff
@@ -0,0 +1,261 @@
+func listKeys(datadir string) ([]string, error) {
+	keystorepath, err := keystoreDir(datadir)
+	if err != nil {
+		return nil, fmt.Errorf("could not get keystore directory: %s", err)
+	}
+
+	files, err := ioutil.ReadDir(keystorepath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read keystore dir: %s", err)
+	}
+
+	keys := []string{}
+
+	for _, f := range files {
+		fmt.Println(f.Name())
+		keys = append(keys, f.Name())
+	}
+
+	return keys, nil
+}
+
+// generateKeypair create a new keypair with the corresponding type and saves it to datadir/keystore/[public key].key
+// in json format encrypted using the specified password
+// it returns the resulting filepath of the new key
+func generateKeypair(keytype, datadir string, password []byte) (string, error) {
+	if password == nil {
+		password = getPassword()
+	}
+
+	if keytype == "" {
+		keytype = utils.Sr25519KeyType
+	}
+
+	var kp crypto.Keypair
+	var err error
+	if keytype == utils.Sr25519KeyType {
+		// generate sr25519 keys
+		kp, err = crypto.GenerateSr25519Keypair()
+		if err != nil {
+			return "", fmt.Errorf("could not generate sr25519 keypair: %s", err)
+		}
+	} else if keytype == utils.Ed25519KeyType {
+		// generate ed25519 keys
+		kp, err = crypto.GenerateEd25519Keypair()
+		if err != nil {
+			return "", fmt.Errorf("could not generate ed25519 keypair: %s", err)
+		}
+	}
+
+	keystorepath, err := keystoreDir(datadir)
+	if err != nil {
+		return "", fmt.Errorf("could not get keystore directory: %s", err)
+	}
+
+	pub := hex.EncodeToString(kp.Public().Encode())
+	fp, err := filepath.Abs(keystorepath + "/" + pub + ".key")
+	if err != nil {
+		return "", fmt.Errorf("invalid filepath: %s", err)
+	}
+
+	file, err := os.OpenFile(fp, os.O_EXCL|os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return "", err
+	}
+
+	defer func() {
+		err = file.Close()
+		if err != nil {
+			log.Error("generate keypair: could not close keystore file")
+		}
+	}()
+
+	err = keystore.EncryptAndWriteToFile(file, kp.Private(), password)
+	if err != nil {
+		return "", fmt.Errorf("could not write key to file: %s", err)
+	}
+
+	log.Info("key generated", "public key", pub, "type", keytype, "file", fp)
+	return fp, nil
+}
+
+// keystoreDir returnns the absolute filepath of the keystore directory given gossamer's datadir
+// by default, it is ~/.gossamer/keystore/
+// otherwise, it is datadir/keystore/
+func keystoreDir(datadir string) (string, error) {
+	// datadir specified, return datadir/keystore as absolute path
+	if datadir != "" {
+		keystorepath, err := filepath.Abs(datadir)
+		if err != nil {
+			return "", err
+		}
+
+		if _, err := os.Stat(keystorepath); os.IsNotExist(err) {
+			err = os.Mkdir(keystorepath, os.ModePerm)
+			if err != nil {
+				return "", err
+			}
+		}
+
+		return keystorepath, nil
+	}
+
+	// datadir not specified, return ~/.gossamer/keystore as absolute path
+	home, err := os.UserHomeDir()
```

Comment:
This can be simplified to: `cfg.DefaultDataDir()`

Reply:
nice, done